File: haskell/problem7.hs

noDivisors []    x                  = True
noDivisors (h:t) x | h*h > x        = True
                   | x `mod` h == 0 = False
                   | otherwise      = noDivisors t x

-- List of prime numbers
primes :: [Int]
primes = 2 : filter (noDivisors primes) [3,5..]

-- Function to check is the number is prime
isPrime = noDivisors primes

prime_10001 = primes !! 10000

File: clojure/problem7.clj
(defn noDivisors [seq x] (cond
                           (empty? seq) true
                           (> (* (first seq) (first seq)) x) true
                           (= 0 (mod x (first seq))) false
                           :else (noDivisors (rest seq) x) ))

(def primes (lazy-cat [2 3] (filter (partial noDivisors primes) (iterate (partial + 2) 5)) ))

(defn prime_10001 [] (nth primes 10000))

(defn main [] (prime_10001))